modules/OFConnectionManager2/module/src/ofconnectionmanager.c

```diff
@@ -0,0 +1,1774 @@
+        return;
+    }
+
+    AIM_LOG_TRACE("Controller stop cxn %s(%p)", cxn->desc, cxn);
+    if (!ind_cxn_is_closed(cxn)) {
+        ind_cxn_stop(cxn);
+    } else {
+        ind_cxn_notify_closed(cxn);
+    }
+}
+
+
+/**
+ * For the specified controller:
+ * - remove all auxiliary connections
+ * - disconnect the main connection
+ */
+void
+controller_disconnect(controller_t *controller)
+{
+    connection_t *main_cxn;
+
+    AIM_LOG_TRACE("Controller disconnect %s(%p)", 
+                  controller->desc, controller);
+
+    remove_aux_cxns(controller, 0);
+
+    main_cxn = controller->cxns[0];
+    if (main_cxn != NULL) {
+        controller_stop_cxn(main_cxn);
+    } else {
+        AIM_LOG_INTERNAL("Controller %s has no main connection",
+                         controller->desc);
+    }
+}
+
+
+/*------------------------------------------------------------
+ * Aux cxn handling
+ *------------------------------------------------------------*/
+
+/**
+ * Remove extra auxiliary connections for this controller, setting the new
+ * number of aux cxns to the specified num_aux.
+ * If num_aux is set to zero, shut down all aux cxns for this controller.
+ */
+static indigo_error_t
+remove_aux_cxns(controller_t *controller, uint32_t num_aux)
+{
+    int idx;
+    uint32_t num_aux_current;
+
+    if (!CONTROLLER_ACTIVE(controller)) {
+        return INDIGO_ERROR_UNKNOWN;
+    }
+
+    if (num_aux == controller->num_aux) {
+        return INDIGO_ERROR_NONE;
+    }
+
+    if (num_aux > controller->num_aux) {
+        AIM_LOG_TRACE("Requested num_aux %d is greater than current aux cxn's "
+                      "for controller %s", num_aux, controller->desc);
+        return INDIGO_ERROR_PARAM;
+    }
+
+    AIM_LOG_VERBOSE("Aux cxns for controller %s changed from %d to %d",
+                    controller->desc, controller->num_aux, num_aux);
+    num_aux_current = controller->num_aux;
+    controller->num_aux = num_aux;
+
+    for (idx = num_aux_current; idx > num_aux; --idx) {
+        controller_stop_cxn(controller->cxns[idx]);
+        controller->cxns[idx] = NULL;
+    }
+
+    return INDIGO_ERROR_NONE;
+}
+
+
+/**
+ * Main and aux connection rules:
+ * If main is closed for any reason, all aux are closed.
+ * If any aux is closed for any error or if the main fails to respond to echo,
+ * all main and aux are closed.
+ * If any aux is closed gracefully, only that aux is closed.
+ * Aux echo timeouts are advisory only; i.e. they do not cause the aux to be
+ * closed.
+ * Only main cxns can be restarted.
+ * Aux cxns are created by controller request and are not restarted when they 
+ * close.
+ */
+
+/**
+ * Set up num_aux auxiliary connections for controller associated
+ * with the specified main connection.
+ *
+ * If num_aux requested is greater than aux cxns we already have,
+ * the extra aux cxns will be removed.
+ */
+int
+ind_cxn_add_aux_cxns(connection_t *main_cxn, uint32_t num_aux)
```

Comment:
This ought to be `ind_cxn_set_aux_cxns`, and `remove_aux_cxns` should be inlined into it.